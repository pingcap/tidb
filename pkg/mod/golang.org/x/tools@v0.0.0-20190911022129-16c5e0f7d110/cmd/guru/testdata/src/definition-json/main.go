package definition

// Tests of 'definition' query, -json output.
// See golang.org/x/tools/cmd/guru/guru_test.go for explanation.
// See main.golden for expected query results.

// TODO(adonovan): test: selection of member of same package defined in another file.

import (
	"lib"
	lib2 "lib"
	"nosuchpkg"
)

func main() {
	var _ int // @definition builtin "int"

	var _ undef           // @definition lexical-undef "undef"
	var x lib.T           // @definition lexical-pkgname "lib"
	f()                   // @definition lexical-func "f"
	print(x)              // @definition lexical-var "x"
	if x := ""; x == "" { // @definition lexical-shadowing "x"
	}

	var _ lib.Type     // @definition qualified-type "Type"
	var _ lib.Func     // @definition qualified-func "Func"
	var _ lib.Var      // @definition qualified-var "Var"
	var _ lib.Const    // @definition qualified-const "Const"
	var _ lib2.Type    // @definition qualified-type-renaming "Type"
	var _ lib.Nonesuch // @definition qualified-nomember "Nonesuch"
	var _ nosuchpkg.T  // @definition qualified-nopkg "nosuchpkg"

	var u U
	print(u.field) // @definition select-field "field"
	u.method()     // @definition select-method "method"
}

func f()

type T struct{ field int }

func (T) method()

type U struct{ T }

type V1 struct {
	W // @definition embedded-other-file "W"
}

type V2 struct {
	*W // @definition embedded-other-file-pointer "W"
}

type V3 struct {
	int // @definition embedded-basic "int"
}

type V4 struct {
	*int // @definition embedded-basic-pointer "int"
}

type V5 struct {
	lib.Type // @definition embedded-other-pkg "Type"
}

type V6 struct {
	T // @definition embedded-same-file "T"
}
