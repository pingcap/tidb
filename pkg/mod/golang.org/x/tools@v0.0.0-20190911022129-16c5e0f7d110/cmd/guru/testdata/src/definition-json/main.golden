-------- @definition builtin --------

Error: int is built in
-------- @definition lexical-undef --------

Error: no object for identifier
-------- @definition lexical-pkgname --------
{
	"objpos": "testdata/src/definition-json/main.go:10:2",
	"desc": "package lib"
}
-------- @definition lexical-func --------
{
	"objpos": "$GOPATH/src/definition-json/main.go:38:6",
	"desc": "func f"
}
-------- @definition lexical-var --------
{
	"objpos": "$GOPATH/src/definition-json/main.go:19:6",
	"desc": "var x"
}
-------- @definition lexical-shadowing --------
{
	"objpos": "$GOPATH/src/definition-json/main.go:22:5",
	"desc": "var x"
}
-------- @definition qualified-type --------
{
	"objpos": "testdata/src/lib/lib.go:3:6",
	"desc": "type lib.Type"
}
-------- @definition qualified-func --------
{
	"objpos": "testdata/src/lib/lib.go:9:6",
	"desc": "func lib.Func"
}
-------- @definition qualified-var --------
{
	"objpos": "testdata/src/lib/lib.go:14:5",
	"desc": "var lib.Var"
}
-------- @definition qualified-const --------
{
	"objpos": "testdata/src/lib/lib.go:12:7",
	"desc": "const lib.Const"
}
-------- @definition qualified-type-renaming --------
{
	"objpos": "testdata/src/lib/lib.go:3:6",
	"desc": "type lib.Type"
}
-------- @definition qualified-nomember --------

Error: couldn't find declaration of Nonesuch in "lib"
-------- @definition qualified-nopkg --------
{
	"objpos": "testdata/src/definition-json/main.go:12:2",
	"desc": "package nosuchpkg"
}
-------- @definition select-field --------
{
	"objpos": "testdata/src/definition-json/main.go:40:16",
	"desc": "field field int"
}
-------- @definition select-method --------
{
	"objpos": "testdata/src/definition-json/main.go:42:10",
	"desc": "func (T).method()"
}
-------- @definition embedded-other-file --------
{
	"objpos": "testdata/src/definition-json/type.go:3:6",
	"desc": "type W int"
}
-------- @definition embedded-other-file-pointer --------
{
	"objpos": "testdata/src/definition-json/type.go:3:6",
	"desc": "type W int"
}
-------- @definition embedded-basic --------

Error: int is built in
-------- @definition embedded-basic-pointer --------

Error: int is built in
-------- @definition embedded-other-pkg --------
{
	"objpos": "testdata/src/lib/lib.go:3:6",
	"desc": "type lib.Type"
}
-------- @definition embedded-same-file --------
{
	"objpos": "$GOPATH/src/definition-json/main.go:40:6",
	"desc": "type T"
}
